compaction.go

```diff
@@ -232,6 +232,42 @@ func (c *compaction) elideRangeTombstone(start, end []byte) bool {
 	return true
 }
 
+// atomicCompactionUnitUpperBound returns the upper boundary of the atomic
+// compaction unit containing the specified sstable (identified by a pointer to
+// its fileMetadata).
+//
+// TODO(peter): This needs to be directly tested.
+func (c *compaction) atomicCompactionUnitUpperBound(f *fileMetadata) []byte {
+	for i := range c.inputs {
+		files := c.inputs[i]
+		for j := range files {
+			if f == &files[j] {
+				upperBound := f.largest.UserKey
+				for j = j + 1; j < len(files); j++ {
+					cur := &files[j-1]
```

Comment:
Oh yeah, it was also clever to operate just on the compaction input files unlike my implementation, which copied `expandInputs` in that it referenced files in `version`. But that was inconvenient because then we had to worry about holding the DB mutex, I think.